### DISCUSSION TITLE
Append middlewares to service only once even when config is included multiple times

### DISCUSSION THREAD

**Discussion by marian13** [marian13/convenient_service#43] [Ideas]
- When the already added config is included multiple times, it appends `middlewares` again and again.

- Check the following example:

  ```bash
  task playground
  ```

  ```ruby
  class TestMiddleware < ConvenientService::Core::MethodChainMiddleware
    def next(...)
      chain.next(...)
    end
  end
  ```

  ```ruby
  module FirstConfig
    def self.included(service_class)
      service_class.include ConvenientService::Standard::Config

      service_class.middlewares :result do
        use TestMiddleware
      end
    end
  end
  ```

  ```ruby
  module SecondConfig
    def self.included(service_class)
      service_class.include ConvenientService::Standard::Config

      service_class.middlewares :result do
        use TestMiddleware
      end
    end
  end
  ```

  ```ruby
  class Service
    include FirstConfig
    include SecondConfig
  end
  ```

  ```ruby
  Service.middlewares(:result).to_a
  # => [ConvenientService::Common::Plugins::NormalizesEnv::Middleware, ConvenientService::Service::Plugins::HasResult::Middleware, ConvenientService::Service::Plugins::HasResultSteps::Middleware, ConvenientService::Common::Plugins::HasCallbacks::Middleware, ConvenientService::Common::Plugins::HasAroundCallbacks::Middleware, ConvenientService::Service::Plugins::RaisesOnDoubleResult::Middleware, ConvenientService::Common::Plugins::CachesReturnValue::Middleware, TestMiddleware, TestMiddleware]
  ```

  ```ruby
  Service.middlewares(:result).to_a.size
  # => 9
  ```

  ```ruby
  class Service
    include FirstConfig
    include SecondConfig
  end
  ```

  ```ruby
  Service.middlewares(:result).to_a.size
  # => 11
  ```

  ```ruby
  class Service
    include FirstConfig
    include SecondConfig
  end
  ```

  ```ruby
  Service.middlewares(:result).to_a.size
  # => 13
  ```
  
- In the future, a lot of the new predefined configs will be created, and some of them may have dependencies on each other.

- As a result, some middleware may be added multiple times unintentionally.

**Complexity: Medium**

**Hard Requirements:**

- Provide a simple way for the end user to avoid such behavior.
- `include Config` signature in `class Service` must be kept as it is, but `module Config` can be modified.
- Corresponding [examples](https://github.com/marian13/convenient_service/tree/main/lib/convenient_service/examples) and [docs](https://marian13.github.io/convenient_service_docs/basics/tldr) should be updated.

**Comment by Olha-30** [comment]
According to my research, it also appends concerns again and again &#129300;

Check this out here:

```ruby
module TestConcern
  include ConvenientService::Support::Concern
end
```
```ruby
module FirstConfig
  def self.included(service_class)
    service_class.include ConvenientService::Standard::Config

    service_class.concerns do
      use TestConcern
    end
  end
end
```
```ruby
module SecondConfig
  def self.included(service_class)
    service_class.include ConvenientService::Standard::Config

    service_class.concerns do
      use TestConcern
    end
  end
end
```
```ruby
class Service
  include FirstConfig
  include SecondConfig
end
```

```ruby
Service.concerns.to_a.count
# => 16
```

```ruby
class Service
  include FirstConfig
  include SecondConfig
end
```
```ruby
Service.concerns.to_a.count
# => 18
```
```ruby
class Service
  include FirstConfig
  include SecondConfig
end
```
```ruby
Service.concerns.to_a.count
# => 20
```

**Reply by marian13** [reply]
> According to my research, it also appends concerns again and again 🤔
> 
> Check this out here:
> 
> ```ruby
> module TestConcern
>   include ConvenientService::Support::Concern
> end
> ```
> 
> ```ruby
> module FirstConfig
>   def self.included(service_class)
>     service_class.include ConvenientService::Standard::Config
> 
>     service_class.concerns do
>       use TestConcern
>     end
>   end
> end
> ```
> 
> ```ruby
> module SecondConfig
>   def self.included(service_class)
>     service_class.include ConvenientService::Standard::Config
> 
>     service_class.concerns do
>       use TestConcern
>     end
>   end
> end
> ```
> 
> ```ruby
> class Service
>   include FirstConfig
>   include SecondConfig
> end
> ```
> 
> ```ruby
> Service.concerns.to_a.count
> # => 16
> ```
> 
> ```ruby
> class Service
>   include FirstConfig
>   include SecondConfig
> end
> ```
> 
> ```ruby
> Service.concerns.to_a.count
> # => 18
> ```
> 
> ```ruby
> class Service
>   include FirstConfig
>   include SecondConfig
> end
> ```
> 
> ```ruby
> Service.concerns.to_a.count
> # => 20
> ```

Thanks, @Olha-30.

Appreciate your help a lot.

Could you also check a case when the config is defined as a concern, please?

For example:

```ruby
module FirstConfig
  include ConvenientService::Support::Concern
  
  included do
    include ConvenientService::Standard::Config

    concerns do
      use TestConcern
    end
  end
end
```

**Reply by Olha-30** [reply]
@marian13, I have checked the case. 
Here is the result:

```ruby 
module TestConcern
  include ConvenientService::Support::Concern
end
```
```ruby 
module FirstConfig
  include ConvenientService::Support::Concern
  
  included do
    include ConvenientService::Standard::Config

    concerns do
      use TestConcern
    end
  end
end
```
```ruby
module SecondConfig
  include ConvenientService::Support::Concern
  
  included do
    include ConvenientService::Standard::Config

    concerns do
      use TestConcern
    end
  end
end
```
```ruby 
class Service
  include FirstConfig
  include SecondConfig
end
```
```ruby
Service.concerns.to_a.count
# => 16
```
```ruby 
class Service
  include FirstConfig
  include SecondConfig
end
```
```ruby
Service.concerns.to_a.count
# => 16
```
As you can see, the amount of concerns does NOT change, no matter how many times I include already-included modules.
So in conclusion, I guess that `included` in concerns already works like `included_once`.  And the behavior of the regular `included` hook in simple modules differs from the concerns `included`, as it does NOT check if the same code was already triggered, and it triggers `middlewares` and `concerns` methods, in our case, again and again. Maybe it was one of the reasons for creating `concerns` in Rails? To improve performance? &#129300; (just assuming &#128516;)

**Reply by Olha-30** [reply]
Here is the mix example, using middlewares and concerns together:
```ruby
class TestMiddleware < ConvenientService::Core::MethodChainMiddleware
  def next(...)
    chain.next(...)
  end
end
```
```ruby
module TestConcern
  include ConvenientService::Support::Concern
end
```
```ruby
module FirstConfig
  include ConvenientService::Support::Concern

  included do
    include ConvenientService::Standard::Config
  
    concerns do
      use TestConcern
    end
 
    middlewares :result do
      use TestMiddleware
    end
  end
end
```
```ruby
module SecondConfig
  include ConvenientService::Support::Concern

  included do
    include ConvenientService::Standard::Config
  
    concerns do
      use TestConcern
    end
 
    middlewares :result do
      use TestMiddleware
    end
  end
end
```
```ruby
class Service
  include FirstConfig
  include SecondConfig
end
```
```ruby
Service.concerns.to_a.count
# => 16
```
```ruby
Service.middlewares(:result).to_a.size
# => 9
```
```ruby
class Service
  include FirstConfig
  include SecondConfig
end
```
```ruby
Service.concerns.to_a.count
# => 16
```
```ruby
Service.middlewares(:result).to_a.size
# => 9
```
The amount of `concerns` and `middlewares` does NOT change &#128516;

**Comment by marian13** [comment]
Thanks, @Olha-30.

Your improvements are merged as a part of https://github.com/marian13/convenient_service/pull/46.
